File: skills/code-architect/references/code-examples/php/Iterator.php
<?php
// Iterator — traverse a collection without exposing its representation.
declare(strict_types=1);

/** @implements IteratorAggregate<int, string> */
final class Ring implements IteratorAggregate
{
    /** @var string[] */
    private array $items = [];

    public function push(string $item): void
    {
        $this->items[] = $item;
    }

    // Custom iterator that walks the items in reverse order.
    public function getIterator(): Iterator
    {
        return new ReverseCursor($this->items);
    }
}

/** @implements Iterator<int, string> */
final class ReverseCursor implements Iterator
{
    private int $pos;

    /** @param string[] $items */
    public function __construct(private array $items)
    {
        $this->pos = count($items) - 1;
    }

    public function current(): string { return $this->items[$this->pos]; }
    public function key(): int { return $this->pos; }
    public function next(): void { $this->pos--; }
    public function rewind(): void { $this->pos = count($this->items) - 1; }
    public function valid(): bool { return $this->pos >= 0; }
}

// Demo — client iterates without knowing the underlying storage.
$ring = new Ring();
$ring->push('one');
$ring->push('two');
$ring->push('three');

foreach ($ring as $index => $value) {
    echo "$index => $value" . PHP_EOL;
}

File: skills/code-architect/references/code-examples/go/iterator.go
// Iterator: traverse the elements of a collection sequentially without
// exposing its underlying representation.
package main

import "fmt"

// Iterator yields elements one at a time.
type Iterator interface {
	HasNext() bool
	Next() string
}

// collection hides its storage behind an iterator factory.
type collection struct {
	items []string
}

func (c *collection) Iterator() Iterator {
	return &cursor{coll: c}
}

// cursor tracks the current position over the collection.
type cursor struct {
	coll *collection
	pos  int
}

func (it *cursor) HasNext() bool { return it.pos < len(it.coll.items) }

func (it *cursor) Next() string {
	item := it.coll.items[it.pos]
	it.pos++
	return item
}

func main() {
	c := &collection{items: []string{"one", "two", "three"}}
	for it := c.Iterator(); it.HasNext(); {
		fmt.Println(it.Next())
	}
}
